### ISSUE TITLE
Would OSGB conversions be appropriate?

### ISSUE THREAD

**Issue by judgej** [thephpleague/geotools#34]
Would OSGB to UTM conversions be appropriate to this library, or is it out of scope?
OSGB is a grid system used in the UK and developed by Ordnance Survey (OS). A conversion of OS data to UTM, and back again, would be very useful.

See http://en.wikipedia.org/wiki/Ordnance_Survey_National_Grid

Just testing the water to see if national grid systems is something you want to support. I've had to do conversions on data like this in the past, and will have some code from years ago, tucked away somewhere (I needed to convert postcode locations from OSGB formats). I expect there would be others from other regions and nations, once that gate is open.

**Comment by coreation** [comment]
Hi @toin0u, really nice work with your geotools library, however I have a question about conversions of coordinates between different reference systems. Since @judgej was working on one for the UK reference system, I thought my comment here would be a good place to start.

So I'm using proj4php since 4 years ago, only using some small conversions between Lambert '73 (or EPSG:31370) or others. However I don't see any possibility of doing this using geotools (is that correct?).

If this is so, I can take the code from proj4php, which is LGPL licensed, and change it to match your library structure. Pretty sure that coordinate conversions fit in this package, definitly since it's a league package :). Hope to hear from you soon!

**Comment by toin0u** [comment]
Hi @coreation, thank you for your kind words :) You're right - it's not possible so you're very welcome to make a PR. I will really appreciate it :)

**Comment by coreation** [comment]
Great! I think I'll put the code in 1 converter class that inherits from your Convert-interface. Creating a point then convert that point into a something else seems like a big change, as it currently doesn't accept any other coordinate system if I'm reading it quickly ;).

So I'll start doing that, see where I end up, I'll mention you in my commits before I make a PR ;)

**Comment by toin0u** [comment]
You read correctly ;) Feel free to ping me anytime! I'm very busy atm but I'll do my best :)

**Comment by judgej** [comment]
Hi @coreation I did start trying to convert proj4 to PHP, and have got some way. That was to convert the OSGB datum to the single datum that geotools uses. IMO that it is the right way to go - to use a separate coordinate conversion library rather than to try to build it into geotools (and of course, duplicating effort and ending up with less flexible tools. Some people are using some parts of that library now, but it still needs more care and attention to finish it off:

https://github.com/academe/proj4php

The basic idea of the library is to be able to pass it some coordinates, tell it what projection and datum it uses, then ask the library to return those some coordinates in any other projection and datum you like. 

There may be something you can use in that library, or better still, help to finish it off. It *is* on my TODO list, but priorities are a constantly shifting thing.

**Comment by coreation** [comment]
Hi @judgej I was thinking about whether or not this would be fitting to use it here. When I took a look at the Convert Interface I was thinking the same thing: "This maybe better off in a separate library".

I see that it's a 5.3 port (or it was when you started) I'd like for this to be 5.4 at minimum, I'll fork your project and take it for a spin. @toin0u keep you posted on this, you can choose then whether or not to pull it into geotools.

**Comment by judgej** [comment]
Yeah, go for it. It was 5.3 because that is where thephpleague libraries were at the time. 5.4 is definitely a good minimum now. 5.5 has some nice features, but it does take time for hosters to catch up.

Let me know if there is anything I can do to help, or anything that needs explaining.

**Comment by coreation** [comment]
Well, (and I've should've seen this coming), there's a port from a project that I was iniitally using, but seemed to lack modern coding ;). This project is called proj4php and initially started by someone in Mexico iirc. This has now been put on [github](https://github.com/muka/ShapeReader). I think this might be a better way to start, and perhaps (I'm sorry to kill your baby) make yours obsolete somewhat?

**Comment by judgej** [comment]
Just for a bit of background on the proj4 conversion, I'm basically taking the JavaScript library and converting it almost line-for-line. Ultimately there may be a better way to organise it and to structure the code, but what it actually *does* is so mathematically complex, I didn't want to have to delve too deeply into the inner workings to rewrite it, because that could be a year-long full-time project on its own.

**Comment by coreation** [comment]
Oh yea, completely agreeing on that ;) That's why I wanted to check what's out there first. I'm not 100% sure, but I think the proj4php port is almost equal to yours, but just more completed in terms of supported coordinate references. The math behind it to actually perform all of the translations is indeed too much for "just a project on the side".

**Comment by judgej** [comment]
https://github.com/muka/proj4php/tree/master/src/proj4php - go for it :-) A bit of namespacing, autoloading and PSR-*, and it could be a nice composer library.

**Comment by coreation** [comment]
Indeed!

**Comment by judgej** [comment]
One thing the JS libraries don't do, but which PHP is probably more suited to, is to drive some of the functionality from data rather than have all those files with a single line of code in. I did try to pull all the data into CSV files, so that perhaps this data can be read and used to initialise the library. There may be some useful snippets in there:

https://github.com/academe/proj4php/tree/master/data

**Comment by coreation** [comment]
Great, thanks for the reference. Currently I'm also working with a new Shp library, I'm converting shape libraries into geojson which needs 1/ a shape reader 2/ geo conversion ;). So I'll look into it as my work progresses.

**Comment by judgej** [comment]
Sorry this has been left hanging for so long. I'm getting back to it now :-)

The OSGB conversion involves a datum shift to WGS84, and that is the complicated bit. I didn't want to put the functionality for that into my OSGB package, because it is a problem that has been solved many times before. The Proj.4 project, written in C and which has been going strong since the early 1980s, does datum shifts and a bunch of projection conversions very well. That has been [ported to JavaScrript](https://github.com/proj4js/proj4js/), and that JavaScript library has been [ported to PHP](https://github.com/proj4php/proj4php).

Now, I have got involved in that latter project to help bring it up to PHP5.4 standards and support composer. That is partly done, but it needs more to be done to shake off a bunch of JavaScript baggage that its conversion has carried with it. Not wanting to do a BC break on its current (large number of) users, I'm experimenting with [a reboot of it here](https://github.com/judgej/Proj4). Where that will go - back into Proj4JS or whether it will be too far out of scope, I don't know - that's the fun of experimentation.

But the end result is that this will be a separate library to handle datum shifts and projections to and from WGS84 (which neither geotools nor its `willdurand/geocoder` dependency handle (quite rightly). So watch both those spaces if you need to use OSGB UK mapping coordinates with this library, or (hopefully) any other type of coordinate system or projection. I'll make sure there are adapters to help all these things plug together.

**Comment by toin0u** [comment]
>  I'll make sure there are adapters to help all these things plug together.

Great work :) I think it's the best alternative :+1:

**Comment by coreation** [comment]
Thanks for the update!
